crates/evm/evm/src/executors/invariant/corpus.rs

```diff
@@ -0,0 +1,447 @@
+            if let Err(err) = write_result {
+                debug!(target: "corpus", %err, "Failed to record call sequence {:?}", &corpus.tx_seq);
+            } else {
+                trace!(
+                    target: "corpus",
+                    "persisted {} inputs for new coverage in {corpus_uuid} corpus",
+                    &corpus.tx_seq.len()
+                );
+            }
+        }
+
+        // This includes reverting txs in the corpus and `can_continue` removes
+        // them. We want this as it is new coverage and may help reach the other branch.
+        self.in_memory_corpus.push(corpus);
+    }
+
+    /// Generates new call sequence from in memory corpus. Evicts oldest corpus mutated more than
+    /// configured max mutations value.
+    pub fn new_sequence(&mut self, test: &InvariantTest) -> eyre::Result<Vec<BasicTxDetails>> {
+        let mut new_seq = vec![];
+        let test_runner = &mut test.execution_data.borrow_mut().branch_runner;
+
+        if !self.in_memory_corpus.is_empty() {
+            let rng = test_runner.rng();
+
+            // Flush oldest corpus mutated more than configured max mutations unless they are
+            // producing new finds more than 1/3 of the time.
+            let should_evict = self.in_memory_corpus.len() > self.corpus_min_size.max(1);
+            if should_evict {
+                if let Some(index) = self.in_memory_corpus.iter().position(|corpus| {
+                    corpus.total_mutations > self.corpus_min_mutations &&
+                        (corpus.new_finds_produced as f64 / corpus.total_mutations as f64) < 0.3
+                }) {
+                    let corpus = self.in_memory_corpus.get(index).unwrap();
+                    let uuid = corpus.uuid;
+                    debug!(target: "corpus", "evict corpus {uuid}");
+
+                    // Flush to disk the seed metadata at the time of eviction.
+                    if let Some(corpus_dir) = &self.corpus_dir {
+                        let eviction_time = SystemTime::now()
+                            .duration_since(UNIX_EPOCH)
+                            .expect("Time went backwards")
+                            .as_secs();
+                        foundry_common::fs::write_json_file(
+                            corpus_dir
+                                .join(format!("{uuid}-{eviction_time}-{METADATA_SUFFIX}"))
+                                .as_path(),
+                            &corpus,
+                        )?
+                    }
+                    // Remove corpus from memory.
+                    self.in_memory_corpus.remove(index);
+                }
+            }
+
+            let corpus_len = self.in_memory_corpus.len();
+            let primary = &self.in_memory_corpus[rng.gen_range(0..corpus_len)];
+            let secondary = &self.in_memory_corpus[rng.gen_range(0..corpus_len)];
+
+            // TODO rounds of mutations on elements?
+            match rng.gen_range(0..=5) {
```

Comment:
can generate an enum here with proptest maybe? to not have the unreachable branch and comments on magic values

Reply:
that's really cool as we can add configurable weights now, added in https://github.com/foundry-rs/foundry/pull/10190/commits/d305fffa516d05a2f790d7a8272618d442678722